### ISSUE TITLE
Built-in function copy() seems to be missing from DaphneDSL

### ISSUE THREAD

**Issue by divjakm** [daphne-eu/daphne#761]
I noticed that the built-in function COPY of the DaphneDSL seems to be missing. The documentation says: 

> `copy(arg:matrix/frame)`
> Creates a copy of arg.
> 

But if I run this code:
```
A = as.matrix<f64>(fill(1,3,3));  // 3x3 matrix filled with value 1
//B = sum(A);    // this works
B = copy(A);	 // this throws error
print(B);
```

I get the following error:

```
[error]: Lowering pipeline error.RewriteToCallKernelOpPass failed with the following message [ no kernels registered for operation `copy` ]
    | Source file -> "/home/divjak/Work/daphne/md/tests/basic/test_copy.daph":9:4
    | 
  9 | B = copy(A);      // this throws error
    |     ^~~
```

And this is the contents of the module_fail.log:

```
module {
  func.func @main() {
    %0 = "daphne.constant"() {value = 1 : si64} : () -> si64
    %1 = "daphne.constant"() {value = "Matrix A:"} : () -> !daphne.String
    %2 = "daphne.constant"() {value = true} : () -> i1
    %3 = "daphne.constant"() {value = false} : () -> i1
    %4 = "daphne.constant"() {value = "Matrix B:"} : () -> !daphne.String
    %5 = "daphne.constant"() {value = 3 : index} : () -> index
    %6 = "daphne.constant"() {value = 105810441194144 : ui64} : () -> ui64
    %7 = "daphne.constant"() {value = 105810441194016 : ui64} : () -> ui64
    %8 = "daphne.constant"() {value = 140737406273624 : ui64} : () -> ui64
    %c0_i32 = arith.constant 0 : i32
    %9 = "daphne.call_kernel"(%8, %7, %6, %c0_i32) {callee = "_createDaphneContext__DaphneContext__uint64_t__uint64_t__uint64_t"} : (ui64, ui64, ui64, i32) -> !daphne.DaphneContext
    %10 = "daphne.createDaphneContext"(%8, %7, %6) : (ui64, ui64, ui64) -> !daphne.DaphneContext
    %c1_i32 = arith.constant 1 : i32
    %11 = "daphne.call_kernel"(%0, %5, %5, %c1_i32, %10) {callee = "_fill__DenseMatrix_int64_t__int64_t__size_t__size_t"} : (si64, index, index, i32, !daphne.DaphneContext) -> !daphne.Matrix<3x3xsi64>
    %12 = "daphne.fill"(%0, %5, %5) : (si64, index, index) -> !daphne.Matrix<3x3xsi64>
    %c2_i32 = arith.constant 2 : i32
    %13 = "daphne.call_kernel"(%12, %c2_i32, %10) {callee = "_cast__DenseMatrix_double__DenseMatrix_int64_t"} : (!daphne.Matrix<3x3xsi64>, i32, !daphne.DaphneContext) -> !daphne.Matrix<3x3xf64>
    %14 = "daphne.cast"(%12) : (!daphne.Matrix<3x3xsi64>) -> !daphne.Matrix<3x3xf64>
    %c3_i32 = arith.constant 3 : i32
    "daphne.call_kernel"(%12, %c3_i32, %10) {callee = "_decRef__Structure"} : (!daphne.Matrix<3x3xsi64>, i32, !daphne.DaphneContext) -> ()
    "daphne.decRef"(%12) : (!daphne.Matrix<3x3xsi64>) -> ()
    %c4_i32 = arith.constant 4 : i32
    "daphne.call_kernel"(%1, %2, %3, %c4_i32, %10) {callee = "_print__char__bool__bool"} : (!daphne.String, i1, i1, i32, !daphne.DaphneContext) -> ()
    "daphne.print"(%1, %2, %3) : (!daphne.String, i1, i1) -> ()
    %c5_i32 = arith.constant 5 : i32
    "daphne.call_kernel"(%14, %2, %3, %c5_i32, %10) {callee = "_print__DenseMatrix_double__bool__bool"} : (!daphne.Matrix<3x3xf64>, i1, i1, i32, !daphne.DaphneContext) -> ()
    "daphne.print"(%14, %2, %3) : (!daphne.Matrix<3x3xf64>, i1, i1) -> ()
    %15 = "daphne.copy"(%14) : (!daphne.Matrix<3x3xf64>) -> !daphne.Matrix<?x?xf64>
    "daphne.decRef"(%14) : (!daphne.Matrix<3x3xf64>) -> ()
    "daphne.print"(%4, %2, %3) : (!daphne.String, i1, i1) -> ()
    "daphne.print"(%15, %2, %3) : (!daphne.Matrix<?x?xf64>, i1, i1) -> ()
    "daphne.decRef"(%15) : (!daphne.Matrix<?x?xf64>) -> ()
    "daphne.destroyDaphneContext"() : () -> ()
    "daphne.return"() : () -> ()
  }
}
```

**Comment by pdamme** [comment]
Thanks for reporting this issue, @divjakm. I can confirm that the `copy()` built-in function in DaphneDSL does not work yet. The reason is that we have never implemented a kernel for it.
- On the one hand, a kernel could easily be implemented following the [guidelines on implementing a built-in kernel](https://daphne-eu.github.io/daphne/development/ImplementBuiltinKernel/) (see also #18).
- On the other hand, thinking about it now, I'm wondering if we really need a `copy()` function in DaphneDSL (see below).

**It would be great if you could let us know what you intend to use the `copy()` function for.** With that information, we could find the best solution for you. *Is your need for a `copy()` function related to the bug described in #625?*

-----
**Background:**
*In fact, we have never needed the copy function so far.* Whenever a DaphneDSL data object is passed to a built-in function or user-defined function, the original data object is never modified. Instead, the output of built-in and user-defined functions is always a fresh data object that you can assign to a new DaphneDSL variable. After that, you still have access to the old value through the old variable name:

```R
X = fill(2.0, 3, 3); # 3x3 matrix filled with twos
Y = sqrt(X);
print(X);            # X is still available
print(Y);            # Y is also available
```
```
DenseMatrix(3x3, double)
2 2 2
2 2 2
2 2 2
DenseMatrix(3x3, double)
1.41421 1.41421 1.41421
1.41421 1.41421 1.41421
1.41421 1.41421 1.41421
```

The only operation where you cannot avoid overwriting a DaphneDSL variable is [left indexing](https://daphne-eu.github.io/daphne/DaphneDSL/LanguageRef/#indexing_1), where you assign new data to a range of an existing data object. But here, the semantics is copy-on-write, i.e., you can make the original data object available by a fresh variable name *before* you do the left-indexing:

```R
X = fill(2.0, 3, 3); # 3x3 matrix filled with twos
Y = X;               # make data object behind X available as Y
X[1, 1] = [0.0];     # set "middle" element to zero
print(X);            # the modified matrix
print(Y);            # the original matrix
```
```
DenseMatrix(3x3, double)
2 2 2
2 0 2
2 2 2
DenseMatrix(3x3, double)
2 2 2
2 2 2
2 2 2
```

Therefore, I currently don't see why one would need a `copy()` function in DaphneDSL. But maybe I'm missing something? Searching for related issues, I found your earlier issue #625 again (which remained unanswered so far, sorry for that, it must have escaped our attention). That is indeed a bug, which we must fix separately.

**Comment by divjakm** [comment]
Thank you for the quick answer and the thorough explanation! I only intended to use the copy function as a workaround for the bug #625 . So I agree that there is no real need for it. But since it was listed in the documentation, I thought it's best to report it.

**Comment by pdamme** [comment]
@divjakm Thanks for the clarification. After the fix of #625 just now, we really don't need a `copy()` function. I have removed it from DaphneDSL and the documentation now.
